enarx-prstate

```diff
@@ -25,19 +26,31 @@ def handle(issue, pr, responsible):
     if needed:
         pr.add_to_assignees(*needed)
 
-# If there are any changes requested, the author is responsible.
+# Look through all PRs that have a changes requested review present.
 for issue in github.search_issues(f"org:enarx is:pr is:public is:open -is:draft review:changes_requested"):
     pr = issue.as_pull_request()
 
-    handle(issue, pr, {pr.user.login})
-
-# Otherwise, requested reviewers and reviewers with dismissed reviews are.
-# This handles both where the author requests re-review and where, after
-# approvals, reviews were dismissed due to a new push.
-for issue in github.search_issues(f"org:enarx is:pr is:public is:open -is:draft -review:changes_requested"):
-    pr = issue.as_pull_request()
-
-    dismissed = {r.user.login for r in pr.get_reviews() if r.state == 'DISMISSED'}
-    requested = {r.login for r in pr.get_review_requests()[0]}
-
-    handle(issue, pr, dismissed | requested)
+    # Collect reviewers whose reviews have been addressed in a set.
+    addressed_reviewers = set()
+
+    # Get the latest review for individual reviewers.
+    latest_reviews = {i.user.id: i for i in pr.get_reviews() if i.state == "CHANGES_REQUESTED"}
+
+    # Parse through the PR's event log to find review requests.
+    for event in pr.get_issue_events():
+        if event.requested_reviewer is not None and event.created_at > latest_reviews[event.requested_reviewer.id].submitted_at:
+            # If the above condition is true, that means the PR author has
+            # re-requested a review from the reviewer after the reviewer's last
+            # CHANGES_REQUESTED review was submitted. That implies the PR is
+            # ready for re-review from that reviewer.
+            addressed_reviewers.add(event.requested_reviewer)
+    
+    # If there are no addressed reviewers, then it's the author's responsibility
+    # to address the reviews that do exist.
+    if len(addressed_reviewers) == 0:
+        handle(issue, pr, {pr.user.login})
+    # Otherwise, it's the responsibility of those whose reviews have been
+    # addressed and those whose reviews were dismissed.
+    else:
+        dismissed = {r.user.login for r in pr.get_reviews() if r.state == 'DISMISSED'}
+        handle(issue, pr, dismissed | addressed_reviewers)
```

Comment:
I think this logic is incorrect. The responsibility falls on reviewers with addressed reviews AND the author (if unaddressed reviews > 0). If it is ready for merge, it should be assigned to `enarxbot`. If there aren't enough reviews in progress, reviewers should be automatically assigned.

Reply:
If a mergeable PR becomes unmergable, it probably needs to be reassigned to the author for a rebase.

Reply:
I think the latter two are best encapsulated as separate tasks. #26 encapsulates "auto assign if not enough reviews" and I've just opened #33 for "assign author if PR becomes unmergeable".